parastell/magnet_coils.py

```diff
@@ -90,292 +144,223 @@ def logger(self):
-
-        # Conditionally extract parameters for circular cross-section
-        if shape == "circle":
-            # Check that list format is correct
-            if len(self._cross_section) == 1:
-                e = ValueError(
-                    "Format of list defining circular cross-section must be\n"
-                    '["circle" (str), radius (float, cm)]'
-                )
-                self._logger.error(e.args[0])
-                raise e
-            elif len(self._cross_section) > 2:
-                w = Warning(
-                    "More than one length dimension has been defined for "
-                    "cross_section. Interpreting the first as the circle's"
-                    'radius; did you mean to use "rectangle"?'
-                )
-                self._logger.warning(w.args[0])
-
-            # Extract parameters
-            mag_len = self._cross_section[1]
-            # Define string to pass to Cubit for cross-section generation
-            shape_str = f"{shape} radius {mag_len}"
-        # Conditinally extract parameters for rectangular cross-section
-        elif shape == "rectangle":
-            # Check that list format is correct
-            if len(self._cross_section) != 3:
-                e = ValueError(
-                    "Format of list defining rectangular cross-section must \n"
-                    'be ["rectangle" (str), width (float, cm), thickness '
-                    "(float, cm)]"
-                )
-                self._logger.error(e.args[0])
-                raise e
-            # Extract parameters
-            width = self._cross_section[1]
-            thickness = self._cross_section[2]
-            # Detemine largest parameter
-            mag_len = max(width, thickness)
-            # Define string to pass to Cubit for cross-section generation
-            shape_str = f"{shape} width {thickness} height {width}"
-        # Otherwise, if input string is neither 'circle' nor 'rectangle',
-        #  raise an exception
-        else:
-            e = ValueError(
-                "Magnet cross-section must be either a circle or rectangle. "
-                "The first entry of the list defining the cross-section must be"
-                " the shape, with the following entries defining the shape"
-                "parameters.\n"
-                "\n"
-                "For a circular cross-section, the list format is\n"
-                '["circle" (str), radius (float, cm)]\n'
-                "\n"
-                "For a rectangular cross-section, the list format is\n"
-                '["rectangle" (str), width (float, cm),'
-                "thickness (float, cm)]"
-            )
-            self._logger.error(e.args[0])
-            raise e
-
-        self.shape = shape
-        self.shape_str = shape_str
-        self.mag_len = mag_len
-
-    def _extract_filaments(self):
-        """Extracts filament data from magnet coil data file.
+    def _extract_filament_data(self):
```

Comment:
This is a good function to review in a software meeting, but we might not want to wait that long.

I think this would be cleaner if you first just converted all the data from the text file into a more convenient numerical format, and then in separate loops or separate functions, downsample the data and compute the tangents.

1. Data file --> list of numpy arrays of coordinates 
2. Downsample numpy arrays, probably with simple array slice notation
3. Compute tangents, ideally in a function that can act on arrays

Reply:
This would also allow you to calculate the tangents at the last minute, rather than carrying them all around until you need them.

Reply:
My intention was to reduce the total amount of information stored, storing only the sampled points while at the same time computing the tangents using the full set of points. But I agree it makes the code a little confusing, and maybe adds to the total amount of information stored at a given time.

Reply:
Also, I was unaware you could use steps when array slicing; very neat!